plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,658 @@
+/*
+ * Copyright (C) 2021 Intel Corporation.
+ * Copyright (C) 2021 Dell Inc.
+ * All rights reserved.
+ *
+ * This software and associated documentation (if any) is furnished
+ * under a license and may only be used or copied in accordance
+ * with the terms of the license.
+ *
+ * This file is provided under a dual MIT/LGPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ * Dell Chooses the MIT license part of Dual MIT/LGPLv2 license agreement.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+ OR MIT
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fwupd-error.h"
+#include "fu-device.h"
+#include "fu-common.h"
+#include "fu-common-version.h"
+
+#include "fu-dell-dock-common.h"
+
+#define GR_USB_INTERFACE_NUMBER	0x0
+#define GR_USB_BLOCK_SIZE	64
+
+/* bmRequest type */
+#define USB_REQ_TYPE_GET_MMIO	0xc0  /* bm Request type */
+#define USB_REQ_TYPE_SET_MMIO	0x40  /* bm Request type */
+
+/* bRequest */
+#define REQ_HUB_GET_MMIO	64
+#define REQ_HUB_SET_MMIO	65
+
+/* wValue*/
+#define MBOX_ACCESS		(1 << 10)
+
+/* wIndex, mailbox register offset */
+/* First 16 registers are Data[0]-Data[15] registers */
+#define MBOX_REG_METADATA	16
+#define MBOX_REG		17 /* no name? */
+
+/* mask for the MBOX_REG register that has no name */
+#define MBOX_ERROR		(1 << 6)/* of the u8 status field */
+#define MBOX_OPVALID		(1 << 7)/* of the u8 status field */
+
+#define MBOX_TIMEOUT		3000
+
+/* HUB operation OP codes */
+#define OP_NVM_WRITE		0x20
+#define OP_NVM_AUTH_WRITE	0x21
+#define OP_NVM_READ		0x22
+#define OP_NVM_SET_OFFSET	0x23
+#define OP_DROM_READ		0x24
+
+/* NVM metadata offset and length fields are in dword units */
+/* note that these won't work for DROM read */
+#define NVM_OFFSET_TO_METADATA(p) ((((p) / 4) & 0x3fffff) << 2) /* bits 23:2 */
+#define NVM_LENGTH_TO_METADATA(p) ((((p) / 4) & 0xf) << 24) /* bits  27:24 */
+
+/* Default length for NVM READ */
+#define NVM_READ_LENGTH		0x224
+#define NVM_VER_OFFSET_MAJOR	0xa
+#define NVM_VER_OFFSET_MINOR	0x9
+#define NVM_VID_OFFSET_MAJOR	0x221
+#define NVM_VID_OFFSET_MINOR	0x220
+#define NVM_PID_OFFSET_MAJOR	0x223
+#define NVM_PID_OFFSET_MINOR	0x222
+
+struct mbox_regx {
+	guint16		opcode;
+	guint8		rsvd;
+	guint8		status;
+} __attribute__ ((packed));
+
+/* 10ms interval when polling for operation complete. */
+struct timespec poll_delay = {.tv_sec = 0, .tv_nsec = 10 * 1000000};
+
+struct _FuDellDockUsb4 {
+	FuUsbDevice	parent_instance;
+	guint32		blob_ver_major_offset;
+	guint32		blob_ver_minor_offset;
+	guint32		blob_vid_major_offset;
+	guint32		blob_vid_minor_offset;
+	guint32		blob_pid_major_offset;
+	guint32		blob_pid_minor_offset;
+	gchar		*nvm_ver;
+	gchar		*nvm_vid;
+	gchar		*nvm_pid;
```

Comment:
I don't see any reason to store these here.  Just locally allocate the string with `g_autofree gchar * = NULL` in the function and then call the functions to add instance ID and version using that.